ton_client/client/src/processing/process_message.rs

```diff
@@ -1,84 +1,98 @@
-//     })
-// }
-//
-// pub async fn process_message(
-//     context: Arc<ClientContext>,
-//     params: ParamsOfProcessMessage,
-// ) -> ApiResult<ResultOfProcessMessage> {
-//     let mut retry_count = 0;
-//     loop {
-//         let (message, expiration_time) =
-//             ensure_message(&context, &params.message, retry_count, &params.callback)?;
-//         let transaction_waiting_state = send_message(
-//             context.clone(),
-//             ParamsOfSendMessage {
-//                 message: message.clone(),
-//                 message_expiration_time: None,
-//                 callback: params.callback.clone(),
-//             },
-//         )
-//         .await?
-//         .transaction_waiting_state;
-//
-//         let result = wait_for_transaction(
-//             context.clone(),
-//             ParamsOfWaitForTransaction {
-//                 message: message.clone(),
-//                 message_expiration_time: expiration_time,
-//                 callback: params.callback.clone(),
-//                 waiting_state: transaction_waiting_state,
-//             },
-//         )
-//         .await?;
-//         match result {
-//             ResultOfWaitForTransaction::Complete(transaction) => {
-//                 emit_event(&context, &params.callback, || {
-//                     MessageProcessingEvent::TransactionReceived {
-//                         transaction: transaction.clone(),
-//                     }
-//                 });
-//                 return Ok(ResultOfProcessMessage { transaction });
-//             }
-//
-//             ResultOfWaitForTransaction::Incomplete(waiting_state) => {}
-//         }
-//         retry_count += 1;
-//     }
-// }
+        // Send
+        let mut processing_state = send_message(
+            context.clone(),
+            ParamsOfSendMessage {
+                message: message.clone(),
+                abi: abi.clone(),
+                callback: params.callback.clone(),
+            },
+        )
+        .await?
+        .processing_state;
+
+        // Monitor network
+        loop {
+            match wait_for_transaction(
+                context.clone(),
+                ParamsOfWaitForTransaction {
+                    message: message.clone(),
+                    callback: params.callback.clone(),
+                    abi: abi.clone(),
+                    processing_state: processing_state.clone(),
+                },
+            )
+            .await
+            {
+                Ok(result) => match result {
+                    ResultOfWaitForTransaction::Complete(output) => {
+                        return Ok(output);
+                    }
+                    ResultOfWaitForTransaction::Incomplete {
+                        processing_state: incomplete_state,
+                        ..
+                    } => {
+                        processing_state = incomplete_state;
+                    }
+                },
+                Err(err) => {
```

Comment:
What if some other error returned? Not `MessageExpired` and `TransactionWaitTimeout`. For example server returned invalid response that couldn't be parsed and will continue to return it? Then this loop will never end. I think here we should retry only network errors. Let's make the rule for defining network errors. And other error we should return to the user

Reply:
is_network_error